### PR TITLE
Feature/354 delete train

### LINKED ISSUE #354: Delete train when it left the sim
## User stories
As a developer I want that trains which left the sim are deleted in order to have an overview which trains are still available

## Acceptance Criteria
- [x] research notifications for a train leaving the sim
- [x] When listing all trains, trains which left the sim shouldn't be available anymore

### PR DESCRIPTION
**Pull request by SaturnHafen**
Fixes #354

Trains are deleted when the id is not available in sumo/traci anymore

## PR checklist

- [x] Acceptance criteria fulfilled
- [x] Additional features are tested
- [ ] ~~Docs (code, wiki & diagrams) updated~~
- [ ] ~~Breaking changes anounced~~
- [ ] ~~Added classes that inherit from `BaseModel` to `src.constants.tables`~~
- [ ] Dev-branch has been merged into local branch to resolve conflicts
- [ ] Tests and linter have passed AFTER local merge
- [ ] Another dev reviewed and approved

### CODE REVIEW

File: tests/wrapper/test_simulation_object_updating_component.py [hunk 1]
```diff
@@ -110,3 +110,15 @@ def test_node_refs(self, configured_component: SimulationObjectUpdatingComponent
                     )
 
                     assert node in (edge.to_node, edge.from_node)
+
+    def test_stale_train_is_removed(
+        self,
+        configured_component: SimulationObjectUpdatingComponent,
+        train: Train,
```
Comment (CR1337) [inline tests/wrapper/test_simulation_object_updating_component.py:117]: Where are you using `train`, `results` and `all_trains` here?
Maybe just remove these.

File: tests/wrapper/conftest.py [hunk 2]
```diff
@@ -11,6 +12,27 @@
 from src.wrapper.train_spawner import TrainSpawner
 
 
+@pytest.fixture
+def results(monkeypatch):
```
Comment (CR1337) [inline tests/wrapper/conftest.py:16]: What are the fixtures `results` and `all_trains` for?

File: tests/wrapper/test_simulation_object_updating_component.py [hunk 3]
```diff
@@ -110,3 +110,15 @@ def test_node_refs(self, configured_component: SimulationObjectUpdatingComponent
...
+        results,
+        all_trains,
+    ):
+        configured_component.next_tick(1)
+        configured_component.next_tick(2)
+
+        assert len(configured_component.trains) == 0
```
Comment (CR1337) [inline tests/wrapper/test_simulation_object_updating_component.py:126]: Probably I just did not understand your code fully. But I could not find where the fixture `configured_component` got any trains added. So are you really testing the deletion of trains here?

**Review by CR1337 — COMMENTED** [review]
Looks good so far. Especially `_remove_stale_vehicles` is very cleanly written in my view. I just got some questions about your test :)

File: tests/wrapper/test_simulation_object_updating_component.py (hunk 1 above)
Comment (SaturnHafen) [reply tests/wrapper/test_simulation_object_updating_component.py:117]: The `train` is the train we are adding to the sim, `all_trains` mocks the request to traci that queries all vehicles in the sim and `results` mocks the `get_all_subscriptions` call to traci

File: tests/wrapper/conftest.py (hunk 2 above)
Comment (SaturnHafen) [reply tests/wrapper/conftest.py:16]: see above

File: tests/wrapper/test_simulation_object_updating_component.py (hunk 3 above)
Comment (SaturnHafen) [reply tests/wrapper/test_simulation_object_updating_component.py:126]: You are correct, fixed that :)

**Review by CR1337 — APPROVED** [review]
It looks good now.

**Pull request merged**